### ISSUE TITLE
What does variational dropout means?

### ISSUE THREAD

**Issue by YingxiaoKong** [ceshine/recurrent-dropout-experiments#1]
I found variational dropout in your python code, I have also read a paper: a theoretically grounded application of dropout in recurrent neural networks, which also mentioned variational dropout. I'm wondering are they the same thing? Have you read this paper before?

I'm looking forward to your reply!

**Comment by ceshine** [comment]
Yes, I've read the paper. This repo is actually used to keep the source code used in this Medium series: [[Learning Note] Dropout in Recurrent Networks — Part 1](https://becominghuman.ai/learning-note-dropout-in-recurrent-networks-part-1-57a9c19a2307).

EDIT: the source code was used specifically for this post, in case you couldn't find it: [[Learning Note] Dropout in Recurrent Networks — Part 3](https://towardsdatascience.com/learning-note-dropout-in-recurrent-networks-part-3-1b161d030cd4).

**Comment by YingxiaoKong** [comment]
Thank you for your reply! I just found them online! It's very helpful! Have you ever considered to combine MC dropout and variational dropout together to see what will happen?

**Comment by ceshine** [comment]
I couldn't remember if I ever published anything combining those two online, but I do have a post comparing MC dropout with quantile regression — [Quantile Regression — Part 2](https://medium.com/the-artificial-impostor/quantile-regression-part-2-6fdbc26b2629). 

There's a paper published by Uber using MC dropout & variational dropout to model uncertainty in time series prediction, in case you're not aware of it.

**Comment by YingxiaoKong** [comment]
Thank you! I will look at the paper later.

在 2019年8月8日，上午1:25，CeShine Lee <notifications@github.com<mailto:notifications@github.com>> 写道：


I couldn't remember if I ever published anything combining those two online, but I do have a post comparing MC dropout with quantile regression ― Quantile Regression ― Part 2<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fmedium.com%2Fthe-artificial-impostor%2Fquantile-regression-part-2-6fdbc26b2629&data=02%7C01%7Cyingxiao.kong%40vanderbilt.edu%7C0f8d9f2b84ec48e23e1b08d71bc93c01%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637008423415201944&sdata=k%2Fs0AKW%2Fmt7VF0vdRWLD3PecNbJB1eZE9fzgOHWb2B8%3D&reserved=0>.

There's a paper published by Uber using MC dropout & variational dropout to model uncertainty in time series prediction, in case you're not aware of it.

―

**Comment by YingxiaoKong** [comment]
HI ceshine,


I just read the paper about varitional dropout again and also looked at your code carefully. I'm still confused of how to apply the same dropout mask at each time step.  In the paper, it mentioned: implementing our approximate inference is identical to implementing dropout in RNNs with the same network units dropped at each time step.

Could you point out in which part you implement this technique? Thank you!

**Comment by ceshine** [comment]
It's implemented in weight_drop.py directly copied from [salesforce/awd-lstm-lm
](https://github.com/salesforce/awd-lstm-lm/blob/master/weight_drop.py). It drops the weights of RNN units randomly and that inherently ensures the dropout mask is the same acorss time steps. This is much faster because we can still use optimized Cudnn RNN units. You can check that repo for more information.

The implementation is a bit outdated, though. Here's an implementation I've found on Google that seems to be more compatible with the latest PyTorch: [torchnlp.nn.weight_drop](https://pytorchnlp.readthedocs.io/en/latest/_modules/torchnlp/nn/weight_drop.html).

**Comment by YingxiaoKong** [comment]
Sorry for a late reply! I'm kind of busy these days and forgot to check my mailbox. I checked your code and found that weight_drop.py are implemented in Pytorch files but not Keras. So in Keras LSTM, how  can we guarantee the same mask at each time step? Thank you!

**Comment by ceshine** [comment]
We don't need to do it ourselves. The `recurrent_dropout` option in Keras implements (or implemented) the variational dropout. I had analyzed the Keras source code in this post — [[Learning Note] Dropout in Recurrent Networks — Part 2](https://towardsdatascience.com/learning-note-dropout-in-recurrent-networks-part-2-f209222481f8). 

Note that it did not use cuDNN LSTM implementation, and therefore very slow in practice. I'm not sure it is still the case (I'd guess so, since last time I checked cuDNN still doesn't support variational dropout in LSTM).

**Comment by YingxiaoKong** [comment]
Hi Ceshine,

Thank you for your reply! Yes recurrent_dropout is pretty slow. I just found it out. I may try py torch later to see how fast it is, or cuDNN LSTM.

在 2019年8月20日，下午10:31，CeShine Lee <notifications@github.com<mailto:notifications@github.com>> 写道：


We don't need to do it ourselves. The recurrent_dropout option in Keras implements (or implemented) the variational dropout. I had analyzed the Keras source code in this post ― [Learning Note] Dropout in Recurrent Networks ― Part 2<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Ftowardsdatascience.com%2Flearning-note-dropout-in-recurrent-networks-part-2-f209222481f8&data=02%7C01%7Cyingxiao.kong%40vanderbilt.edu%7C077f2853af994f95367608d725e80295%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637019550718713803&sdata=M0aobJaUuD9UjQxZa2TJmBbxrM1U6qF9z1yZtN731eU%3D&reserved=0>.

Note that it did not use cuDNN LSTM implementation, and therefore very slow in practice. I'm not sure it is still the case (I'd guess so, since last time I checked cuDNN still doesn't support variational dropout in LSTM).

―
